### PR TITLE
DEV-3086: Fix resume download invalid range

### PR DESCRIPTION
**Pull request by jonhenrik13**
This pull request adds comprehensive tests for partial download file handling in the file manager, focusing on security, correctness, and cleanup behavior. It introduces tests to ensure partial downloads are correctly named, safely stored, and properly cleaned up, especially when digests or file names are malformed or manipulated. Additionally, it updates an existing test to verify that leftover partial downloads from previous digests are always removed.

Partial download file path and directory handling:

* Added `Test_GetPartialDownloadFilePath` to verify that partial download paths include the digest, remain in the correct directory, handle malformed digests safely, and keep file names within system limits.
* Added `Test_openPrivatePartialDownloadDirectory` to ensure partial download directories are created with correct permissions and that symlinked directories are rejected for security.

Partial download and cleanup logic:

* Added `Test_downloadMetadataCompare_RemovesPartialDownloadsWithOtherDigest` to confirm that stale partial downloads from previous digests are removed, while unrelated files are preserved.
* Updated `Test_ResumeDownload` to explicitly create and clean up stale partial downloads for different contents, and to check that both the stale file and its directory are deleted after a successful download. [[1]](diffhunk://#diff-03937e26ca8d55627eab922e46e8e8ba89d1415ac9a447b09664b3993d283d14L436-R440) [[2]](diffhunk://#diff-03937e26ca8d55627eab922e46e8e8ba89d1415ac9a447b09664b3993d283d14R471-R496)

Edge case and security tests:

* Added tests for edge cases such as file names exceeding `NAME_MAX`, zero-size metadata rejection, and handling of symlinked partial files to prevent security issues.

Test utility improvements:

* Added helper functions like `sha256Hex` and improved imports to support new test cases. [[1]](diffhunk://#diff-4b2871e236b0edc0e9b5b7d7e8e43c853a79f79c2c5e69419094e9425965bae8R21-R32) [[2]](diffhunk://#diff-4b2871e236b0edc0e9b5b7d7e8e43c853a79f79c2c5e69419094e9425965bae8R344-R563)

### CODE REVIEW

File: app/configuration/file_manager.go [hunk 1]
```diff
@@ -198,21 +226,69 @@ func (srv *Service) downloadMetadataCompare(ctx context.Context, label, src, dst
...
+	if err = fd.Chmod(fileManagerDefaultFilePermission); err != nil {
+		return false, fmt.Errorf("error setting permissions on %s: %w", tmpDst, err)
 	}
 
 	if err = os.Rename(tmpDst, dst); err != nil {
 		return false, fmt.Errorf("error renaming file %s to %s: %w", tmpDst, dst, err)
 	}
+	if err = removePartialDownloadDirectory(tmpDst); err != nil {
+		return false, err
+	}
```
[the remark this answers is not in the document]
Comment (jonhenrik13) [reply app/configuration/file_manager.go]: Fixed

File: app/configuration/file_manager.go [hunk 2]
```diff
@@ -198,21 +226,69 @@ func (srv *Service) downloadMetadataCompare(ctx context.Context, label, src, dst
...
+
+	return finalizePartialDownload(ctx, label, src, dst, tmpDst, fileMetadata, fileCreateData)
+}
+
+// finalizePartialDownload verifies a fully downloaded partial file and moves it to its destination.
+//
+// tmpDst is opened with O_NOFOLLOW so a symlink placed at the predictable partial download path
+// cannot be verified through its target and then have that symlink (rather than a regular file)
+// installed at dst by os.Rename (CWE-59).
```
[the remark this answers is not in the document]
Comment (jonhenrik13) [reply app/configuration/file_manager.go]: Addressed.

File: app/configuration/file_manager.go [hunk 3]
```diff
@@ -140,85 +149,205 @@ func (srv *Service) downloadMetadataCompare(ctx context.Context, label, src, dst
...
+	offset    int64
+}
+
+// preparePartialDownload creates and validates the agent-owned directory before inspecting its entry.
+func preparePartialDownload(dst string, fileMetadata *FileMetadata, fileCreateData *fileCreateData) (*partialDownload, error) {
+	if err := makeDirectories(dst, fileManagerDefaultDirectoryPermission, fileCreateData.uid, fileCreateData.gid); err != nil {
+		return nil, err
+	}
+
+	path := GetPartialDownloadFilePath(dst, fileMetadata.Digest())
```
[the remark this answers is not in the document]
Comment (jonhenrik13) [reply app/configuration/file_manager.go:205]: We assume legacy downloads being completed.

File: app/configuration/file_manager.go [hunk 4]
```diff
@@ -140,85 +149,218 @@ func (srv *Service) downloadMetadataCompare(ctx context.Context, label, src, dst
...
+	}
+	if fileMetadata.Size == 0 && partial.offset > 0 {
+		partialReady, readyErr := isFileReady(partial.path, fileMetadata)
+		if readyErr != nil {
+			return false, readyErr
+		}
+		if partialReady {
+			return srv.finalizeDownloadedFile(ctx, label, src, dst, partial, fileMetadata, fileCreateData)
+		}
+	}
```
[the remark this answers is not in the document]
Comment (jonhenrik13) [reply app/configuration/file_manager.go]: we want to treat zero file size in the metadata as an error
